### DISCUSSION TITLE
My thoughts after the first month of use

### DISCUSSION THREAD

**Discussion by Kha1r0** [nikitabobko/AeroSpace#645] [feature-ideas]
Hey there!
I've switched to aerospace about a month ago and for my workflow, it's a big upgrade from the classic Sequoia experience or even Software like Rectangle.
Especially the straightforward config was very easy to get into.
There's two key faults i see in aerospace and want to spark a discussion about, especially 2. is very important to me.
# 1. Localized keyboard layouts
The default config is completely unusable with the german macOS keyboard layout.
This is because it binds to option + 56789 and L, which in the german macOS keyboard are already bound to [, ], |, {, } and @.

So for my config, i had to unbind the workspaces 56789, so i can use the brackets.
I also had to switch the navigation keys from HJKL to UIOP, so option + L is not blocked for the @ symbol.

Getting to this point as a beginner is not quite beginner friendly. It could be suggested to provide localized layouts or promote a place to show off own configs so beginners can more easily get started. Once i got the hang of it though, it wasn't a problem anymore.

# 2. Using aerospace alongside macOS spaces, not as a replacement
As [Josean Martinez](https://www.youtube.com/@joseanmartinez) correctly mentions in [his introduction](https://youtu.be/-FoWClVHG5g?t=200), aerospace reimplements workspaces, because there are limitations with the built in Spaces, especially in terms of keyboard shortcuts and animation speed.

The problem is, that it misses some of Mission Control and macOS spaces's most important benifits.
1. Visual Overview
	- When in mission control, it is very easy to see a) which windows are open b) in which space, to get a grand perspective over the spaces.
	- Aerospace's workspaces completely lack this functionality. (Going through all the workspaces is not a suitable replacement.)
	- (Yes, enabling "Group Windows by application" makes it possible to graphically see every single open window, but that's not helpful since it doesn't show them in the context of their workspace, e.g. one browser instance for each workspace.)
1. Mouse/Trackpad Gestures and their Speed
	 - Speaking on behalf of people, who sometimes lie in bed with their laptop, _i don't want_ to use the keyboard to navigate 100% of the time.
	 - The perfect window manager needs to be good at managing windows with the keyboard, as well as the mouse. The user has to have the opportunity to choose.
	 - What aerospace lacks are the great and partly pretty fast Trackpad/Mouse gestures that macOS has for switching between workspaces and getting an overview (e.g. three finger swipe sideways, three finger swipe up).

## So, what's the solution?
1. Completely discard Mouse Navigation, which will only make 100% keyboard navigation people happy, or:
2. give users the opportunity to either turn aerospace workspaces off completely, making only use of the aerospaces great tiling opportunities (Which would solve the situation for me already) or
3. Implement a mode, where aerospace workspaces work together with macOS spaces. e.g. only numbered workspaces corresponding to the macOS numbered workspaces.

Now, can't we already just use macOS Spaces and just stay on aerospace workspace 1 all the time? That would be nice, but no, we cannot. Let me give you an example.

Let's say I'm on a macOS Space and I'm on aerospace workspace 3. Both are empty:

<img width="1512" alt="Screenshot 2024-10-29 at 22 41 20" src="https://github.com/user-attachments/assets/3e139acd-1456-405e-921d-8b12f5f3b41d">

When i open a window, it tiles correctly as expected.

<img width="1512" alt="Screenshot 2024-10-29 at 22 41 36" src="https://github.com/user-attachments/assets/b139e497-b41b-4fcc-aea4-86bb712b99ae">

But if i now open a new macOS Space, that also belongs to workspace 3 and open another window, the newly opened window tiles **WITH THE WINDOW ON ANOTHER macOS SPACE.**

<img width="1512" alt="Screenshot 2024-10-29 at 22 41 56" src="https://github.com/user-attachments/assets/393c9eca-f270-49a9-b86c-bb61a21a3ee1">

Going back to the first space, we see, that it tiles with the other space, making it completely impossible to use macOS spaces, while aerospace is installed.

<img width="1512" alt="Screenshot 2024-10-29 at 22 42 01" src="https://github.com/user-attachments/assets/6de0c730-d484-449f-8695-8d6cb8accd98">

Besides, we will never get rid of macOS spaces anyways. Things like full screen Browser videos will always create a macOS Space.
Users need a way to turn off aerospace workspaces, if they prefer to use aerospace only for window tiling, not for workspace management. Because macOS is already very good at virtual desktop management.
Thanks for reading through this. I'm eager to hear about your opinion.
Cheers!

**Comment by Fanatic17** [comment]
I apologize for not reading all of it, I'm a fellow user of aerospace but I wanted to tell you a couple of things:

- Localization: yes i had the same problem too (italian keyboards, even though I switch between italian an US layout i need a reliable method to switch spaces in both) and i solved it assigning ctrl instead of option as the modifier to switch space and shift + ctrl to move a window to a monitor. I just wish i could configure reliably also qwerty or something with ctrl but it is used as a shortcut in many apps
- Turning aerospace off: in the menubar you can pause it or quit and it's basically the same thing or maybe I'm missing the point; keep in mind I only have one mac workspace since i began using aerospace. 

I'll probably make my own post with my thoughts about this, this was inspirational

**Reply by Kha1r0** [reply]
I'm not talking about turning aerospace completely, but disabling aerospace workspaces, and keep using the tiling window functionality.

**Reply by EvansMatthew97** [reply]
Interestingly, I find the most value in the opposite. macOS spaces are slow and cumbersome, and AeroSpace makes moving windows and switching workspaces near instant by comparison. I don't need the tiling so much - if every app opened in floating by default I would still use AeroSpace.

If you want just the tiling window functionality, there are other tools out there like Rectangle Pro which can do some pretty advanced things.

**Comment by Neo3477** [comment]
I have used BetterTouchTool to switch between workspaces:
3 finger swipe to the left: /opt/homebrew/bin/aerospace workspace next
3 finger swipe to the right: /opt/homebrew/bin/aerospace workspace prev

**Reply by Kha1r0** [reply]
Thats a great idea. But what do you do when you have a browser video maximized and it creates another space?
Also, the Visual Overview remains to be solved.

**Reply by olirogers** [reply]
If you're coming here in 2026, there's this project - [MediosZ/SwipeAeroSpace](https://github.com/MediosZ/SwipeAeroSpace) - that allows mapping of the 3 finger swipe for aerospace.

**Comment by shishkin** [comment]
I wonder if aerospace list-* queries could be used to build 3rd party workspace visualizer. I understand, the aerospace author is not keen on GUI bells and whistles, which is fair.

**Reply by EvansMatthew97** [reply]
It looks like the current best/easiest solution is SketchyBar, which is [documented in the Goodies](https://nikitabobko.github.io/AeroSpace/goodies#show-aerospace-workspaces-in-sketchybar).

That's what I've done and it works pretty well. It's a bit finicky to setup - there's no perfect default setup. Usually you'll be copying someone else's bar config and scrapping the parts you don't need/want. I would personally love a lightweight GUI companion that's a bit more responsive and more feature-rich than SketchyBar without taking a weekend to get it running. 

I find the visual indicator absolutely necessary to keep track of where anything is. Mnemonics only got me so far. Maybe a skill issue.

I would guess it's best someone create a separate repository for a companion app written in something like Swift. I don't know if there are low level bindings into aerospace.
